cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1853,70 +1926,247 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
+    _column_index_truncate_length(options.get_column_index_truncate_length()),
+    _kv_meta(options.get_key_value_metadata()),
+    _single_write_mode(mode == SingleWriteMode::YES),
+    _out_sink(std::move(sinks))
+{
+  if (options.get_metadata()) {
+    _table_meta = std::make_unique<table_input_metadata>(*options.get_metadata());
+  }
+  init_state();
+}
+
+writer::impl::~impl() { close(); }
+
+void writer::impl::init_state()
+{
+  _current_chunk_offset.resize(_out_sink.size());
+  // Write file header
+  file_header_s fhdr;
+  fhdr.magic = parquet_magic;
+  for (auto& sink : _out_sink) {
+    sink->host_write(&fhdr, sizeof(fhdr));
+  }
+  std::fill_n(_current_chunk_offset.begin(), _current_chunk_offset.size(), sizeof(file_header_s));
+}
+
+void writer::impl::write(table_view const& input, std::vector<partition_info> const& partitions)
+{
+  _last_write_successful = false;
+  CUDF_EXPECTS(not _closed, "Data has already been flushed to out and closed");
+
+  if (not _table_meta) { _table_meta = std::make_unique<table_input_metadata>(input); }
+  fill_table_meta(_table_meta, input);
+
+  // All kinds of memory allocation and data compressions/encoding are performed here.
+  // If any error occurs, such as out-of-memory exception, the internal state of the current
+  // writer is still intact.
+  [[maybe_unused]] auto [updated_agg_meta,
+                         chunks,
+                         pages,
+                         global_rowgroup_base,
+                         first_rg_in_part,
+                         batch_list,
+                         rg_to_part,
+                         uncomp_bfr,   // unused, but contains data for later write to sink
+                         comp_bfr,     // unused, but contains data for later write to sink
+                         col_idx_bfr,  // unused, but contains data for later write to sink
+                         write_buff,
+                         num_columns] = [&] {
+    try {
+      return convert_table_to_parquet_data(input,
+                                           partitions,
+                                           *_table_meta,
+                                           _agg_meta,
+                                           _kv_meta,
+                                           _max_page_fragment_size,
+                                           _max_row_group_size,
+                                           _max_page_size_bytes,
+                                           _max_row_group_rows,
+                                           _max_page_size_rows,
+                                           _column_index_truncate_length,
+                                           _stats_granularity,
+                                           _compression,
+                                           _dict_policy,
+                                           _max_dictionary_size,
+                                           _single_write_mode,
```

Comment:
if you don't, please change `_single_write_mode` to type `SingleWriteMode`. `bool` is just a step back here, since we already have a nice enum.